File: go/solutions/0345-reverse-vowels-of-a-string.go
package solutions

func ReverseVowels(s string) string {
	i, j := 0, len(s)-1
	runes := []rune(s)
	for i < j {
		if !isVowel(runes[i]) {
			i++
			continue
		}

		if !isVowel(runes[j]) {
			j--
			continue
		}

		runes[i], runes[j] = runes[j], runes[i]
		i++
		j--
	}

	return string(runes)
}

func isVowel(c rune) bool {
	return (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' ||
		c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U')
}

File: js/0345-reverse-vowels-of-a-string.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
  var i = 0,
    j = s.length - 1;

  var isVowel = function (c) {
    return (
      c == 'a' ||
      c == 'e' ||
      c == 'i' ||
      c == 'o' ||
      c == 'u' ||
      c == 'A' ||
      c == 'E' ||
      c == 'I' ||
      c == 'O' ||
      c == 'U'
    );
  };

  s = s.split("");
  while (i < j) {
    if (!isVowel(s[i])) {
      i++;
      continue;
    }

    if (!isVowel(s[j])) {
      j--;
      continue;
    }

    var temp = s[i];
    s[i] = s[j];
    s[j] = temp;

    i++;
    j--;
  }

  return s.join("");
};

module.exports = { reverseVowels };
